pyomo/dae/set_utils.py

```diff
@@ -0,0 +1,173 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright 2017 National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and 
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain 
+#  rights in this software.
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+from collections import Counter
+
+
+def is_explicitly_indexed_by(comp, s):
+    """
+    Returns True if component comp is directly indexed by set s.
+    """
+    if not comp.is_indexed():
+        return False
+    n = comp.index_set().dimen
+    if n == 1:
+        if comp.index_set() is s:
+            return True
+        else:
+            return False
+    elif n >= 2:
+        if s in set(comp.index_set().set_tuple):
+            # set_tuple must be converted to a python:set so a different
+            # pyomo:set with the same elements will not be conflated.
+            # This works because pyomo:Set is hashable.
+            return True
+        else:
+            return False
+
+
+def is_implicitly_indexed_by(comp, s, stop_at=None):
+    """
+    Returns True if any of comp's parent blocks are indexed by s.
+    Works by recursively checking parent blocks.
+
+    If block stop_at (or its parent_component) is provided, function 
+    will return False if stop_at is reached, regardless of whether 
+    stop_at is indexed by s. Meant to be an "upper bound" for blocks 
+    to check, like a flowsheet.
+    """
+    parent = comp.parent_block()
+
+    # Stop when top-level block has been reached
+    while not (parent is None):
+        # If we have reached our stopping point, quit.
+        if parent is stop_at:
+            return False
+
+        # Look at the potentially-indexed block containing our component.
+        parent = parent.parent_component()
+        # Check again for stopping point in case IndexedBlock object was used.
+        if parent is stop_at:
+            return False
+
+        # Check potentially-indexed block for index s.
+        if is_explicitly_indexed_by(parent, s):
+            return True
+        # Continue up the tree, checking the parent block of our potentially-
+        # indexed block, which I would like to assume contains the BlockData
+        # we started from.
+        else:
+            parent = parent.parent_block()
+    # Return False if top-level block was reached.
+    return False
+
+
+def get_index_set_except(comp, *sets):
```

Comment:
This function seems to duplicate what can be done with a `Reference`.  Can you elaborate more on the expected usage here?  In particular, the use of the `index_getter` function?

Reply:
This function is heavily inspired by `get_index_information` in `dae.misc`. Its intended use is for components with the same number of indexing sets, some of which are identical. The non-identical sets are intended but not required to be continuous, differing by bounds, discretization, etc. 

The goal is, given some elements for the non-identical sets, to find every index of the components for which these non-identical sets (time and space usually) take on the values specified. The usage is:

```python
if is_explicitly_indexed_by(var1, time):
    info = get_index_set_except(var1, time)
    non_time_set = info['set_except']
    index_getter = info['index_getter']
    for partial_index in non_time_set:
        new_index = index_getter(partial_index, new_time_value)
        # do something with new_index
```

`index_getter` inserts values for the sets that have been excluded into partial indices to form a new index with the new values for these sets.

This is useful for
- initializing dynamic models to values from a corresponding steady state model
- initializing dynamic models to their initial conditions 
- copying values between two dynamic (or PDE) models of different discretizations

You could probably do this with references, something like:
```python
for index in Reference(var1[t0, ...]): # requires knowledge of where time is in var's index
    # index is a partial index excluding time
    # need to insert new time value into index
    var2[new_index].set_value(var1[old_index])
```
 I think I remember trying, running into a snag somewhere, and going with this approach instead as it was more similar to what I already had implemented.